### ISSUE TITLE
WiFi connecting to (not configured) Open AP.

### ISSUE THREAD

**Issue by bighippo999** [esphome/issues#3508]
### The problem

I'm having a problem with one of my ESP32 boards. It's new and in a new location but is almost a copy of the running configs from my other boards (just the name changed, and 1 sensor/pin commented out).
The problem is , it was connected absolutely fine, but then it's gone off the network (I know sounds simple and stupid). I've reboot it and it didn't come back. So I plugged it into the Laptop to check the logs of what's going on.
```
[D][wifi:388]: Found networks:
[I][wifi:432]: - 'BTWi-fi' [redacted]▂▄▆█
[D][wifi:433]:     Channel: 1
[D][wifi:434]:     RSSI: -91 dB
[I][wifi:432]: - 'BTWi-fi' [redacted]▂▄▆█
[D][wifi:433]:     Channel: 11
[D][wifi:434]:     RSSI: -93 dB
[I][wifi:432]: - 'SkyNet AI - D' [redacted]▂▄▆█
[D][wifi:433]:     Channel: 1
[D][wifi:434]:     RSSI: -84 dB
[I][wifi:432]: - 'SkyNet AI - S' [redacted]▂▄▆█
[D][wifi:433]:     Channel: 1
[D][wifi:434]:     RSSI: -85 dB
[I][wifi:432]: - 'SKY683LK' [redacted]▂▄▆█
[D][wifi:433]:     Channel: 11
[D][wifi:434]:     RSSI: -89 dB
[I][wifi:432]: - 'Sapphire' [redacted]▂▄▆█
[D][wifi:433]:     Channel: 1
[D][wifi:434]:     RSSI: -91 dB
[I][wifi:432]: - '' (HIDDEN) [redacted]▂▄▆█
[D][wifi:433]:     Channel: 1
[D][wifi:434]:     RSSI: -92 dB
[I][wifi:432]: - '' (HIDDEN) [redacted]▂▄▆█
[D][wifi:433]:     Channel: 11
[D][wifi:434]:     RSSI: -92 dB
[I][wifi:248]: WiFi Connecting to 'BTWi-fi'...
[I][wifi:505]: WiFi Connected!
[C][wifi:353]:   Local MAC: 0C:XX:15:XX:XX:BC
[C][wifi:354]:   SSID: [redacted]
[C][wifi:355]:   IP Address: 100.100.2.123
[C][wifi:357]:   BSSID: [redacted]
[C][wifi:358]:   Hostname: 'esphome-web-XXXXbc'
[C][wifi:360]:   Signal strength: -87 dB ▂▄▆█
[C][wifi:364]:   Channel: 1
[C][wifi:365]:   Subnet: 255.255.0.0
[C][wifi:366]:   Gateway: 100.100.0.1
[C][wifi:367]:   DNS1: 192.168.22.22
[C][wifi:368]:   DNS2: 192.168.22.23
[D][wifi:514]: Disabling AP...
[C][web_server:092]: Setting up web server...
[C][ota:085]: Over-The-Air Updates:
[C][ota:086]:   Address: esphome-web-XXXXbc.local:3232
[C][ota:089]:   Using Password.
[W][ota:095]: Last Boot was an unhandled reset, will proceed to safe mode in 8 restarts
[C][api:025]: Setting up Home Assistant API server...
[I][app:062]: setup() finished successfully!
[D][text_sensor:067]: 'Bathroom Sensor Latest Scan Results': Sending state 'BTWi-fi: -91dB
BTWi-fi: -93dB
SkyNet AI - D: -84dB
SkyNet AI - S: -85dB
SKY683LK: -89dB
Sapphire: -91dB
```
The problem is, this isn't my wifi, it's not configured to use it. I use:
```
wifi:
  networks:
    - ssid: !secret wifi_ssid
    - password: !secret wifi_password
    - priority: 10
    - ssid: !secret wifi_ssid_2
    - password: !secret wifi_password_2
    - priority: 5
    - ssid: !secret wifi_ssid_3
    - password: !secret wifi_password_3
    - priority: 2
```
with the SSIDs and passwords in the secrets file. SkyNet AI - D is the first ssid. I have 2 others for other locations that aren't in range here. I think the BTWi-fi is an open point (from previous experience connecting to BT points on my phone).

Any idea why it would be connecting to this point that's not configured for it to use? and how I stop it? it's not the strongest signal -91db & -93db. My point is -84db and the one that's priority 10. priority 5 and 2 are miles away.

### Which version of ESPHome has the issue?

On this particular board the 1 version before ESPHome version 2022.8.0. Haven't update as it's not on WiFi.

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2022.8.6

### What platform are you using?

ESP32

### Board

esp32dev

### Component causing the issue

WiFi

### Example YAML snippet

```yaml
wifi:
  networks:
    - ssid: !secret wifi_ssid
    - password: !secret wifi_password
    - priority: 10
    - ssid: !secret wifi_ssid_2
    - password: !secret wifi_password_2
    - priority: 5
    - ssid: !secret wifi_ssid_3
    - password: !secret wifi_password_3
    - priority: 2
```

### Anything in the logs that might be useful for us?

```txt
[D][wifi:388]: Found networks:
[I][wifi:432]: - 'BTWi-fi' [redacted]▂▄▆█
[D][wifi:433]:     Channel: 1
[D][wifi:434]:     RSSI: -91 dB
[I][wifi:432]: - 'BTWi-fi' [redacted]▂▄▆█
[D][wifi:433]:     Channel: 11
[D][wifi:434]:     RSSI: -93 dB
[I][wifi:432]: - 'SkyNet AI - D' [redacted]▂▄▆█
[D][wifi:433]:     Channel: 1
[D][wifi:434]:     RSSI: -84 dB
[I][wifi:432]: - 'SkyNet AI - S' [redacted]▂▄▆█
[D][wifi:433]:     Channel: 1
[D][wifi:434]:     RSSI: -85 dB
[I][wifi:432]: - 'SKY683LK' [redacted]▂▄▆█
[D][wifi:433]:     Channel: 11
[D][wifi:434]:     RSSI: -89 dB
[I][wifi:432]: - 'Sapphire' [redacted]▂▄▆█
[D][wifi:433]:     Channel: 1
[D][wifi:434]:     RSSI: -91 dB
[I][wifi:432]: - '' (HIDDEN) [redacted]▂▄▆█
[D][wifi:433]:     Channel: 1
[D][wifi:434]:     RSSI: -92 dB
[I][wifi:432]: - '' (HIDDEN) [redacted]▂▄▆█
[D][wifi:433]:     Channel: 11
[D][wifi:434]:     RSSI: -92 dB
[I][wifi:248]: WiFi Connecting to 'BTWi-fi'...
[I][wifi:505]: WiFi Connected!
[C][wifi:353]:   Local MAC: 0C:XX:15:XX:XX:BC
[C][wifi:354]:   SSID: [redacted]
[C][wifi:355]:   IP Address: 100.100.2.123
[C][wifi:357]:   BSSID: [redacted]
[C][wifi:358]:   Hostname: 'esphome-web-XXXXbc'
[C][wifi:360]:   Signal strength: -87 dB ▂▄▆█
[C][wifi:364]:   Channel: 1
[C][wifi:365]:   Subnet: 255.255.0.0
[C][wifi:366]:   Gateway: 100.100.0.1
[C][wifi:367]:   DNS1: 192.168.22.22
[C][wifi:368]:   DNS2: 192.168.22.23
[D][wifi:514]: Disabling AP...
[C][web_server:092]: Setting up web server...
[C][ota:085]: Over-The-Air Updates:
[C][ota:086]:   Address: esphome-web-XXXXbc.local:3232
[C][ota:089]:   Using Password.
[W][ota:095]: Last Boot was an unhandled reset, will proceed to safe mode in 8 restarts
[C][api:025]: Setting up Home Assistant API server...
[I][app:062]: setup() finished successfully!
[D][text_sensor:067]: 'Bathroom Sensor Latest Scan Results': Sending state 'BTWi-fi: -91dB
BTWi-fi: -93dB
SkyNet AI - D: -84dB
SkyNet AI - S: -85dB
SKY683LK: -89dB
Sapphire: -91dB
```

**Comment by bighippo999** [comment]
Update:
I've confirmed the board does connect to an open BTWi-fi when it shouldn't be. I set my phone as an open hotspot 'BTWi-fi' and reset the device. It connect straight to my phone, ip 192.168.43.25. going to that I can access the esp32 boards webpage.
I did try to upload new firmware (all other devices have been upgrade after HA/esphome update) using the OTA update on the page. but I'm getting 'Wrong Magic Byte' I've searched that, but the main suggestion was it's the wrong file. I'm definitely uploading the bin file (HA>ESPHome>Device>Install>Manual Download>Modern format'). I'm wondering if it's failing because I have OTA password set, but the board web server doesn't ask me for this password. Long shot I know, but I used this same method for updating other boards when I first started out without an issue, but I since set an OTA password and just update from HA wirelessly now.

I'm going to test again setting my phone as 'CTWi-fi', I have a theory it may just be going through the available WiFi in alphabetical order. Will update with the outcome. but this has got me thinking, I've seen posts from others having wifi drops/issues, I wonder if they have any open points near them and it could be jumping over. Mine was running for hours without an issue, happily connected to HA, then dropped out and ever since appears to be offline but in actual fact is connected to BTWi-fi that's not mine.

**Comment by bighippo999** [comment]
Theory somewhat busted :(
I changed the mobile hotspot name to CTWi-fi, while I had the esp board connected to my laptop and could see the logs.
I saw the wifi fail the beacon, then reset the wifi, it scanned and connected back to CTWi-fi. but when I reset the board, it connected to BTWi-fi. So maybe it held the station mac address between changing it and just connected back, I don't know.
I also tried setting the mobile hotspot to 'SkyNet AI - D' still an open point, and coming up with a signal -50, compared to BT's -91. Reset the board and it still connect to the BT point with a much lower signal. I also set the hotspot with the password for SkyNet AI - D, as that's the one the board has the config for and it would now be a really good signal. It still connect to the open BTWi-fi.

I'm at a loss on how to get it to ignore the BT Access Point and actually connect to what it's supposed to. I think I'm only having issues with this one board because the BT point must be closer to it and just on the very edge of being detectable. I have another board in the bedroom about 1.5m away and that's not detecting BTWi-fi as a point, so I may just be lucky that the others can't see it and are all behaving.

As always, any thoughts on where I go from here would be welcome and I'm happy to carry out other tests if anyone can think of anything to try.

**Comment by bighippo999** [comment]
I've had a chance this weekend to do a little more testing.
Hopefully someone can pickup this issue and resolve it as I think it's potentially a decent problem affecting many without knowing about it.

As it stands, I have 4 ESP32 boards now all failing to connect and tonight while trying to connect to some GoSund Plugs that I flashed a few weeks ago (they've been unplugged for a week or so) none of them would connect to the Wifi (I think that's a router problem), they all create an AP.
I can see my test board that was configured with the 3 wifi networks, was trying to connect to the GoSund Fallback AP, even though it is passworded and 100% not configured as something any of the ESP32 boards should attempt to connect to.

I update the yaml of my test board to use just 1 wifi ssid and password instead of the 3 under networks.
As soon as it boot, it connect to the network correctly, even though there are plenty of other points both open and passworded that it was previously trying to connect to.
I then set my phone hotspot to be BTWi-fi and open auth. reset my test board, it ignored my phone and connect to the configured ssid.

So my guess is the way in which wifi>networks is handled. It seems to ignore the configuration and the priority, instead connecting to the first point that will accept it. In the case of BTWi-fi it being an open network, the boards can connect. when it tries a network that has a password, it fails, scans, and tries again or tries another. I could see my test board trying to connect to the different GoSund fallback APs. I'm not sure what criteria it's using to choose though. It doesn't appear to be Signal strength as worse signals are tried first. It's not Name, while BT would be close to the top a hotspot starting with A wasn't tried and instead it still connect to BTWi-fi.

Bottom line: If your board isn't connecting to wifi and you've configured it with multiple SSIDs/networks, try stripping it back to just one ssid and password under the wifi not network. That what seems to be working for me atm.

I'll update this if I figure out anything else. As above, I hope someone can pickup this issue. Unless you can view the logs, you'd have no idea why it's not connecting to wifi, but technically it may be, just not the wifi you configured/expect it to be connected to. This does open it to the possibility of being reprogrammed if it joins another network and you don't have OTA updates disabled/passworded.

Would love some feedback if anyone can at least replicated this issue.
